### PR TITLE
Portable chemical dispensers (for chemists, to offer new style of gameplay)

### PR DESCRIPTION
**Pull request by LordVollkorn**
## About The Pull Request
Adds a portable chemical dispenser to the game, a bulky item that is worn in the belt slot and dispenses and mixes chemicals like a normal dispenser, but not as effective.
It runs on power cells and does not recharge them like the normal dispenser does, requiring the user to switch the power cells out with a screwdriver. It cannot be upgraded with parts such as matter bins or manipulators, keeping it at a permanent low-tier level. **To avoid usage through people other than chemists, it is offered in the chemical autodrobe and cannot be printed or ordered.** It comes with it's own unique sprites, but is almost entirely based upon pre-existing code.

## Why It's Good For The Game
I see chemists plumbing and mixing every round, taking up to one hour to get a decent setup or to make all the useful pills. And of those potentially hundreds of pills they made, only 10-15 are used by the crew. Because before anyone gets to eat a pill, the self-healing-self-respiration super virus, the amazing-super-nanites, the paramedicing mediborgs with sensors, the normal paramedics, the very-bored-and-eagerly-waiting wound-tending medical doctors and the Omnizine-Super-Botany-Food heal those hurt crew members 5 times over. Frustrated and bored Chemists run out of gameplay during the round and go to sleep. Let us bring the dispenser to where it is needed and therefore give Chemists a new approach to the game.

Instead of wearing a chemical pill bag in the belt slot, you can equip this portable dispenser. It gives chemists the possibility to become more mobile, mixing chemicals on the fly and giving incentives to create recipes for useful chemicals that are requested often. A chemist who does not have to hastily run across the whole station in order to quickly produce anti-radiation chemicals is a more efficient chemist. He might not be able to heat, grind or make pills with it, but if it fits in a beaker, you got yourself some chems on the go and some extended gameplay that goes beyond your pharmacy.

**We believe it is balanced because:**
- It does not automatically recharge
- It is not upgradable (no Oil, Ammonia etc.)
- You cannot produce anything that requires a heater (Chlorine Trifluoride, Meth, Ash)
- You cannot selectively add or remove single chemicals (which you could with a ChemMaster)
- You cannot create pills, patches or bottles
- It is bulky and takes up the belt slot (where the chemistry bag would go)
- It can't be used to produce pure versions of the best chemicals (they will be diluted with leftover reagents) so the usage of lower-tier medicine is encouraged
- It can't be produced through normal means, so it is limited (and if it is destroyed, it is gone for good)
- Self-caused explosions such as Potassium+Water are just as easy when asking a Chemist for some Potassium and the bartender for a glass of water (The portable dispenser does not increase griefing or suicide bombing)
- There is nothing you could produce/cause that you couldn't have done with a normal dispenser and a few beakers
- The portable dispenser with it's limited range of recipes is unsuitable for killing people quickly and effectively using any chemical method (In any case, the target either has time to run away or you die yourself)

## Changelog
:cl: 
pcd
Jay Smith (Thanks to Michael Trauf)
add: Portable Chemical Dispenser
imageadd: Unique sprites for the Portable Chemical Dispenser
code: Added Portable Chemical Dispenser to Chemistry AutoDrobe
/:cl:

### CODE REVIEW

**Comment by Pumpkinoe** [comment]
wouldn't this be better in the Nanomed tho since jobdrobes usually don't have functional stuff

**Comment by TerraGS** [comment]
Look at this nerd wearing a chemistry set

**Comment by nianjiilical** [comment]
if someone wearing this gets exploded the machine should break and spew random chemical clouds/foam everywhere

**Comment by MrDoomBringer** [comment]
this should really be available from techwebs tbpqh

**Comment by optimumtact** [comment]
there's no way I can merge this, it's way too powerful

**Comment by ArcaneMusic** [comment]
First off, really solid first PR codewise.
Thing is , chemical dispensers are in a pretty bad place right now, as there is a framework on hack.dm that involves getting rid of chemical dispensers altogether (See any of the documents about merging botany and chemistry as a combined way to produce chemicals). As it stands though, yeah, this would be wayyy too powerful with all the solo-dispenser chems you can make

**Comment by MrMelbert** [comment]
this seems like it has some pretty redonkulous potential with pyrotechnics - a few that come to mind is being able to macro and instantly make smoke (or even smoke + any toxin) / flash powder, or phlogiston, emp, or clf3 + pyrosium, etc. get yourself a firesuit or atmos hardsuit and you can just walk around dispensing fire from your belt.

**Comment by LordVollkorn** [comment]
**I try to answer to some comments, perhaps it will clear up some confusion:**

> 
> 
> wouldn't this be better in the Nanomed tho since jobdrobes usually don't have functional stuff
I was discussing where to put it and the conclusion was, that it should be at a place that only/mostly chemists can access, so that no random assistants go grab it while they are in medbay

> 
> 
> there's no way I can merge this, it's way too powerful

Is there anything you would suggest that could balance it more? I had a discord discussion on what disadvantages it should have, so its not too powerful.
Currently we think it is balanced because:
- It does not automatically recharge
- It is not upgradable (no Oil, Ammonia etc.)
- It does not heat, divide or make chemicals into pills/patches
- It is bulky and takes up the belt slot
- It can't be used to produce pure versions the best chemicals (they will be diluted with leftover reagents)
- It can't be produced infinitly
- Self-Explosions such as Potassium+Water are just as easy when taking two beakers from a normal dispensers and mixing them on the go (does not increase griefing or suicide bombing)

I can't really think of any way this item could create unfair, powerful advantages, because it is inferior to the real dispensers and offers no possibility that you couldn't have done before.



> 
> 
> this seems like it has some pretty redonkulous potential with pyrotechnics - a few that come to mind is being able to macro and instantly make smoke (or even smoke + any toxin) / flash powder, or phlogiston, emp, or clf3 + pyrosium, etc. get yourself a firesuit or atmos hardsuit and you can just walk around dispensing fire from your belt.

That is true, however you could already achieve this with spray bottles and a chemistry bag full of beakers that you keep emptying in each other. As for traitors, I think they have more powerful weaponry available in their uplinks, like the flamethrower or simply guns.


> 
> 
> First off, really solid first PR codewise.
> Thing is , chemical dispensers are in a pretty bad place right now, as there is a framework on hack.md that involves getting rid of chemical dispensers altogether (See any of the documents about merging botany and chemistry as a combined way to produce chemicals). As it stands though, yeah, this would be wayyy too powerful with all the solo-dispenser chems you can make
> 
> 
> Could you clean this up a bit with some early returns? This method is a bit ugly.

It's my first time coding byond and even using git, I will try to get the code to acceptable levels.
I heared about the merging of botany and chemistry and I am not really in favor of it, since **I believe (as long-term chemist) that taking things away from chemistry is the wrong approach**. I see chemists plumbing and mixing every round, taking up to one hour to get a decent setup or to make all the useful pills. And of those potentially hundreds of pills they made, only 10-15 are used by the crew. Because before **anyone** gets to eat a pill, the self-healing-self-respiration super virus, the amazing-super-nanites, the paramedicing mediborgs with sensors, the normal paramedics, the very-bored-and-eagerly-awaiting-patients medical doctors and the Omnizine-Super-Botany-Food heals those hurt crew members 5 times over, -so all Chemists are asked to do is make Pentetic acid or Inacusiate. And I really think taking away more from chemistry is therefore the wrong way, and instead we should give chemistry this new item, so they don't give up the round after 30-60 minutes of feeling relatively useless.

**Comment by trollbreeder** [comment]
Wow. That is a beautiful sprite. Amazing

**Comment by gitbirb** [comment]
> > Question: How many units of chemicals can they make on a full charge?
> 
> Currently the energy efficiency is at 0.1, so at power cell with a charge of 1000 will provide 100 chemicals. You need to keep in mind, that some chemicals require other chemicals to be made first, so 100 is not alot in my opinion.
> However, since you can put in all kinds of power cells, I have been considering to adjust the energy efficiency so you get even less out of a power cell (balancing and such)

The main reason people bring inducers to chem dispensers is because you have to deconstruct a chem dispensor to recharge a battery. Will it be less difficult to do so with the porta, thus making the entire idea of that one downside basically useless? 

From what i see though this is not a sidegrade, but an upgrade. Maybe that's good or bad, i dunno. You should probably make it premium in the vendor though if your gonna put it in a vendor, max it out at like 2 and have it need like 2k credits or something to buy, somethin' like that.

> this should really be available from techwebs tbpqh
Only if it came from the bepis, otherwise mass-printing of these could very well be OP, given the fact you'd be giving portable chem dispensors to 90% of the med team

**Comment by LordVollkorn** [comment]
> From what i see though this is not a sidegrade, but an upgrade.
Right, I didn't see it from that point of view yet. It is easier to change power cells and I would argue that this still might be alright because the power cell is the only thing that you can take out.
And with those rapid parts exchange devices from science, the big dispenser can be upgraded to tier 4 from outside the room, while the portable one is always stuck at the lowest tier with the lowest efficiency

File: code/game/objects/items/devices/portable_chem_dispenser.dm [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+		if(slot_flags == ITEM_SLOT_BELT)
+			if(user.get_item_by_slot(ITEM_SLOT_BELT) == src)
+				if(cell)
+					ui_interact(user)
+				else
+					to_chat(user, "<span class='warning'>It has no power cell installed!</span>")
+			else
+				to_chat(user, "<span class='warning'>You must strap the portable chemical dispenser's belt on to handle it properly!</span>")
+		return
+	return ..()
```
[the remark this answers is not in the document]
Comment (LordVollkorn) [reply code/game/objects/items/devices/portable_chem_dispenser.dm:166]: Sorry, I am not very experienced in coding byond,
are you referring to get rid of the return in line 166 and instead use returns per if/else statement?

[the remark this answers is not in the document]
Comment (LordVollkorn) [reply code/game/objects/items/devices/portable_chem_dispenser.dm:166]: Thank you, I will clean it up along with making other tweaks that I get from the feedback here

**Comment by MrMelbert** [comment]
if this gets closed for being unbalanced, maybe consider making it admin only and giving it to the medical ERT? i feel like it would fit well on the position of a field medic arriving to stabilize the situation

**Comment by LordVollkorn** [comment]
> 
> 
> if this gets closed for being unbalanced, maybe consider making it admin only and giving it to the medical ERT? i feel like it would fit well on the position of a field medic arriving to stabilize the situation

Thank you for this suggestion, but I really hope it won't be closed due to being unbalanced. We had discussions on discord on how this item could probably cause unbalanced gameplay or other problems and we really, with all honesty, can't think of any way this portable dispenser could be too powerful. I believe that many overestimate the things chemistry can do. You can make someone go to sleep with chloral hydrate - after 20 seconds. You can light someone on fire - and watch them run away and extinguish themselves. You can shoot them with toxins - and wait 10 seconds for them to notice and run to medbay.
Or you can get a gun.

You can't even make the best chemicals with this because you don't have a heater. No meth bombs, no chlorine trifluoride, no ash.
And if somebody REALLY insists, I can also remove stable plasma from the reagent list, so you can't produce any phlogiston (plasma fires) anymore either

File: code/game/objects/items/devices/portable_chem_dispenser.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (lordpidey) [reply code/game/objects/items/devices/portable_chem_dispenser.dm:167]: Not what I meant.

```

if(loc != user)
	return
if(!(slot_flags & ITEM_SLOT_BELT))
	return
if(user.get_item_by_slot(ITEM_SLOT_BELT) != src)
	to_chat(user, "<span class='warning'>You must strap the portable chemical dispenser's belt on to handle it properly!</span>")
	return
if(cell)
    ui_interact(user)
else
    to_chat(user, "<span class='warning'>It has no power cell installed!</span>")
```

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]
I hate chem dispensers so I am very much against a pocket one BUT I will allow this if it is techweb bound AND requires a BS Anomaly (because it pulls out chems from thin air). I cannot promise it will stay in the game though.

Absolutely not going to be merged as a vendor item, premium or otherwise.

**Comment by Mycah142** [comment]
> AND requires a BS Anomaly
> 


Ah yes, yet another item that requires the same exact anomaly core as every other core-required item. Lets just ignore the fact that there's what, four or five literally useless anomaly cores that do NOTHING for ANYTHING, and just keep slapping Bluespace Core requirements onto things.

Make it a Flux Anomaly core, because explaining any sort of "generation" of something "From nothing" being Bluespace related isn't even really bluespace related. Bluespace is translocation. Flux is closer to alchemy, or generating an unrelated object from surrounding objects nearly imperceptibly. Plus, it's a chemistry related item, it's closer to Alchemy, and Flux sounds the most related.

Also, real sick of seeing Bluespace being the hand-wave for fucking everything.

**Comment by LordVollkorn** [comment]
> I hate chem dispensers so I am very much against a pocket one BUT I will allow this if it is techweb bound AND requires a BS Anomaly (because it pulls out chems from thin air). I cannot promise it will stay in the game though.
> Absolutely not going to be merged as a vendor item, premium or otherwise.

I am absolutely for balancing and the techweb would be ok, but isn't the BS anomaly a bit excessive? It makes chemicals out of electricity like the normal dispenser

**Comment by antropod** [comment]
Maybe you just think its overpowered? We can add it to the game and balance it later, after we see some use cases. Look at how content is added to other games: they make it slightly overpowered, everyone starts to use it, then they balance it (Dota is one example)

**Comment by antropod** [comment]
Also I haven't seen any arguments why its overpowered from maitainers. Please provide justification for your point of view. I play chemist and doctor, and I see how it's not OP.

**Comment by LordVollkorn** [comment]
> 
> 
> Also I haven't seen any arguments why its overpowered from maitainers. Please provide justification for your point of view. I play chemist and doctor, and I see how it's not OP.

Yes, I would also appreciate some clarification, because I spent a very long time/asked a lot of people for their input in order to write this PR's description with all it's arguments and balancing ideas

File: code/game/objects/items/devices/portable_chem_dispenser.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (LordVollkorn) [reply code/game/objects/items/devices/portable_chem_dispenser.dm]: Thank you lordpidey, I inserted your version and it looks much better.
The old one is currently commented out, I forgot to remove it but I will do that when I got clarification on the mergers requirements and do the final push

**Comment by MrMelbert** [comment]
> You cannot produce anything that requires a heater
I'll point out that you can just use pyrosium in your beaker to heat up anything. With a little practice, you can make meth entirely in the chem dispenser. 
> It is not upgradable (no Oil, Ammonia etc.)
Also this isn't really a downside, most chemists have the upgraded formulas hardwired into their brain anyways

**Comment by Capsandi** [comment]
As the [holy design doc dictates](https://hackmd.io/@tgstation/BkNXSBfuI#Desired-State-of-Reagent-Pipeline) is soon to remove electricity-chemical dispensers, and require base chemicals be mostly sourced from hydroponics. This falls under maintainer design flows, which means it's the sanctioned plan for where the chemistry department is going. If you would like to propose a different plan for the chemistry department, you can do that under the section labeled "non-maintainer design flows".
If your goal with this is to make chemists a more mobile role, you could integrate that idea into something that abides by and works with the sanctioned plan, like a chemical dispenser belt which needs to be refilled manually, but allows chemists to move around the station to get to where those chemicals are being created.

If this was added to the game now, as is, it would be removed with the rest of the magical chemical makers.  Again, if you would like to propose a different plan for chemistry, edit the "non-maintainer design flows" section of the design doc outlining what it would be and why it would be more fun than the sanctioned plan.

**Comment by MrMelbert** [comment]
> Or you could just use a lighter, it's faster than using the chemical heater when not making meth, and doesn't use up space in the beaker.

I was just demonstrating it's possible to make any chem (even meth) without the beaker even leaving the dispenser.

**Comment by antropod** [comment]
That's why it was in the chemdrobe in the first place, so only chemist can have easy access to it. Chemist can still make all those reactions you talking about by using beakers with ingredients. And by putting it into techweb and locking it under anomaly core you basically make it available to science only.

**Comment by ExcessiveUseOfCobblestone** [comment]
> Make it a Flux Anomaly core, because explaining any sort of "generation" of something "From nothing" being Bluespace related isn't even really bluespace related. Bluespace is translocation. Flux is closer to alchemy, or generating an unrelated object from surrounding objects nearly imperceptibly. Plus, it's a chemistry related item, it's closer to Alchemy, and Flux sounds the most related.

Sure, it can be flux too. I was just thinking it would be a nice item to compete with the BoH.

Hope you're feeling better after writing your post.

**Comment by Mycah142** [comment]
> Hope you're feeling better after writing that post.

I am a salt golem.

**Comment by lordpidey** [comment]
Honestly:

That design doc you are upholding as the bible is terrible and awful.

Remember:  SS13 is fun to play.  If Engineering isn't doing their job when it's most needed, you play SS13 on a station with no lights or power.  If the chef isn't doing his job when it's most needed, you play SS13 with a starving character.  If the janitor dies, you play on a messy station.  If there's no medical when it's most needed, you don't play SS13, you stare at a ghost.

**Comment by ExcessiveUseOfCobblestone** [comment]
> 
> 
> Maybe you just think its overpowered? We can add it to the game and balance it later, after we see some use cases. Look at how content is added to other games: they make it slightly overpowered, everyone starts to use it, then they balance it (Dota is one example)

Nerfing is such a pita because you have the same group of people claim it actually isnt as strong despite me being able to (ab)use these everytime I play the role.

I prefer it undertuned then slowly crept up to desirability.

**Comment by antropod** [comment]
Even if chemistry changes, it does not make this item useless. The ability to make chemicals is the same as in stationary chem dispenser. You change one, you change the other. When changes come, you can make chem dispenser a tool to help you make chemicals, like a convenient beaker storage, so you don't have all your 20 reagents laying around on a table in beakers. And portable one will be the same, just maybe less efficient.

**Comment by lordpidey** [comment]
@ExcessiveUseOfCobblestone 

How would you feel about making it unchargable?  No cell replacement, once it's used up, it's done.

**Comment by anconfuzedrock** [comment]
this would've been overpowered in 2016, but currently most chems are shit.

**Comment by LordVollkorn** [comment]
I would, very much, like to avoid adding an anomaly core
Here are some other suggestions that might be acceptable (please not all at once):
- Removing the base reagents Stable Plasma, Mercury and Sulfur to reduce the number of recipes (mostly harmful)
- Reducing energy efficiency further
- Testmerging in the current state and determining in practical gameplay how it fits in
- Adding an ID check for chemistry access in order to use it

**Comment by antropod** [comment]
It was intended to be a gameplay-choice item: use chemistry bag and premake stuff, or make things on-the-go.
That's why it was belt-only and bulky, so you can't have both. And chemdrobe vending was for only chemists to have easy access to it.
Making it anomaly core + techfab gives access to everyone, especially science, which has too much power right now anyways.

**Comment by Mycah142** [comment]
I can't wait for the chem belt to be printable by Science so when they get the Anomaly Core themselves, they just make their own chemistry belt with the omnilathe they build themselves which they get through the tech tree regardless.

Chemistry belt? More like science-makes-drugs belt.

**Comment by MrMelbert** [comment]
> Removing the base reagents Stable Plasma, Mercury and Sulfur to reduce the number of recipes (mostly harmful)

This is actually really practical and prevents the really stupid stuff like macroing flash powder for an instant 5 second hardstun, or clf3/pyrosium.

Unrelated, but are you aware the current chem dispenser is already labeled 'portable'? You might want to change the name of the board to prevent confusion.

image

**Comment by LordVollkorn** [comment]
The last push is my current suggestion, we still have those options from my last comment too

**Comment by ExcessiveUseOfCobblestone** [comment]
Acting like the limited anomaly balance with a medical techfab somehow makes it more accessible to others than a vending machine that is allowed by literally everyone if you have the cash and often out of the purview of other medical staff (aka easier to break in) is slightly concerning. 

If you're expecting chemists to pick it up literally every round so people won't be able to nab it, the "alternative style of play" is not so alternative. I know for a fact this item would be rushed by botany players if chem wasn't picking up the tab. It's a buff to either the chemist if he is making items on the go that dont need high resolution edits a chem master requires, or a HUGE BUFF to literally every other job (well the 2 lucky people that get them) that now has access to the **entirety** of chems arsenal/progression/whatever you want to call it without any work whatsoever except spending dosh, doubly so for people like the chef (maybe on some maps?)/botanist who have access to a chemmaster. Might even have one in the public garden or something.

At best, this throws in another  (2 actually) magic machines with another set of upfront free resources (may/may not be renewable) without any adjustment for the stationary ones. It's definitely a creative PR, but this further integrates perhaps the worst aspect of the job right now which is that nearly the entire department is bound to this stupid machine.

This machine is debilitating for people who want to expand the department to something that is actually worth not fucking off for or making meth and i'm really not interested in further integrating it into our game. The more I think about it the less I want to see it in every round (or even in the game really).

I'll give this some more thought but I don't know if I can see it being good for the game, especially as a common item.

**Pull request closed without merging**